chia/wallet/wallet_coin_store.py

```diff
@@ -62,58 +62,65 @@ async def create(cls, wrapper: DBWrapper):
         await self.db_connection.execute("CREATE INDEX IF NOT EXISTS wallet_id on coin_record(wallet_id)")
 
         await self.db_connection.commit()
-        self.coin_record_cache = dict()
-        self.coin_wallet_record_cache = {}
-        all_coins = await self.get_all_coins()
-        for coin_record in all_coins:
-            self.coin_record_cache[coin_record.coin.name()] = coin_record
-
-        self.wallet_cache_lock = asyncio.Lock()
+        self.coin_record_cache = {}
+        self.unspent_coin_wallet_cache = {}
+        await self.rebuild_wallet_cache()
         return self
 
     async def _clear_database(self):
         cursor = await self.db_connection.execute("DELETE FROM coin_record")
         await cursor.close()
         await self.db_connection.commit()
 
+    async def rebuild_wallet_cache(self):
+        # First update all coins that were reorged, then re-add coin_records
+        all_coins = await self.get_all_coins()
+        self.unspent_coin_wallet_cache = {}
+        self.coin_record_cache = {}
+        for coin_record in all_coins:
+            name = coin_record.name()
+            self.coin_record_cache[name] = coin_record
+            if coin_record.spent is False:
+                if coin_record.wallet_id not in self.unspent_coin_wallet_cache:
+                    self.unspent_coin_wallet_cache[coin_record.wallet_id] = {}
+                self.unspent_coin_wallet_cache[coin_record.wallet_id][name] = coin_record
+
     # Store CoinRecord in DB and ram cache
     async def add_coin_record(self, record: WalletCoinRecord) -> None:
         # update wallet cache
+        name = record.name()
+        self.coin_record_cache[name] = record
+        if record.wallet_id in self.unspent_coin_wallet_cache:
+            if record.spent and name in self.unspent_coin_wallet_cache[record.wallet_id]:
+                self.unspent_coin_wallet_cache[record.wallet_id].pop(name)
+            if not record.spent:
+                self.unspent_coin_wallet_cache[record.wallet_id][name] = record
+        else:
+            if not record.spent:
+                self.unspent_coin_wallet_cache[record.wallet_id] = {}
+                self.unspent_coin_wallet_cache[record.wallet_id][name] = record
 
-        await self.wallet_cache_lock.acquire()
```

Comment:
Is it ok to remove this lock?

Reply:
yes, the cache only gets changed in other lock. (block)